test/classes-additional.coffee

```diff
@@ -0,0 +1,301 @@
+
+    classA = """
+    class A 
+      constructor: (@drink) ->
+      make: -> "Making a #{@drink}"
+    """
+
+    # This throws since during compile super isn't in a known state.
+    # This incorrect syntax is technically valid, but is actually ES6 syntax.
+    # I suspect this will be a common error 
+    throwsD = """
+    class D extends A 
+    D::make = (@flavor) -> super.make() + " with #{@flavor}"
+    d = new D('Machiato')
+    """
+    throws -> CoffeeScript.run classA + throwsD, bare: yes
+
+
+test "bound functions without super", ->
+  # Bound function with @
+  # This should throw on compile, since bound 
+  # constructors are illegal
+  ctorA = (@drink) =>
+    
+  class A 
+    constructor: ctorA
+    make: => 
+      "Making a #{@drink}"
+  ok (new A('Machiato')).make() isnt "Making a Machiato" 
+
+  # extended bound function, extending fails too.
+  class B extends A
+  b = new B('Machiato')
+  ok (new B('Machiato')).make() isnt "Making a Machiato" 
+
+test "super in a bound function", ->
+  class A 
+    constructor: (@drink) ->
+    make: -> "Making a #{@drink}"
+  
+  class B extends A
+    make: (@flavor) =>
+      super + " with #{@flavor}"
+
+  b = new B('Machiato')
+  eq b.make('vanilla'),  "Making a Machiato with vanilla" 
+
+  # super in a bound function in a bound function
+  class C extends A
+    make: (@flavor) =>
+      func = () =>
+        super + " with #{@flavor}"
+      func()
+
+  c = new C('Machiato')
+  eq c.make('vanilla'), "Making a Machiato with vanilla" 
+
+  # bound function in a constructor
+  class D extends A
+    constructor: (drink) ->
+      super(drink)
+      x = =>
+        eq @drink,  "Machiato" 
+      x()
+  d = new D('Machiato')
+  eq d.make(),  "Making a Machiato" 
+
+# duplicate 
+test "super in a try/catch", ->
+  classA = """
+  class A 
+    constructor: (param) ->
+      throw "" unless param
+  """
+
+  throwsB = """
+  class B extends A
+      constructor: ->
+        try
+          super
+  """
+
+  throwsC = """
+  ctor = ->
+    try
+      super
+  
+  class C extends A
+      constructor: ctor
+  """
+  throws -> CoffeeScript.run classA + throwsB, bare: yes
+  throws -> CoffeeScript.run classA + throwsC, bare: yes
+     
+test "mixed ES6 and CS6 classes with a four-level inheritance chain", ->
+  # Extended test 
+  # ES2015+ class interoperability
+
+  ```
+  class Base {
+    constructor (greeting) {
+      this.greeting = greeting || 'hi';
+    }
+    func (string) {
+      return 'zero/' + string;
+    }
+    static  staticFunc (string) {
+      return 'static/' + string;
+    }
+  }
+  ```
+
+  class FirstChild extends Base
+    func: (string) ->
+      super('one/') + string
+
+
+  ```
+  class SecondChild extends FirstChild {
+    func (string) {
+      super.func('two/' + string);
+    }
+  }
+  ```
+
+  thirdCtor = ->
+    @array = [1, 2, 3]
+
+  class ThirdChild extends SecondChild
+    constructor: -> 
```

Comment:
This *should* work as just `constructor: thirdCtor`.

**Edit** actually `thirdCtor` would have to be updated to return `this` or `undefined`.